File: apps/go/design-pattern/command/command.go
// Package command demonstrates the Command pattern in Go.
// Commands are represented as an interface with Execute, Undo, and Description.
package command

import (
	"os"
	"path/filepath"
)

// Command is the interface for all commands.
type Command interface {
	Execute() error
	Undo() error
	Description() string
}

// CreateFileCommand creates a file with given content.
type CreateFileCommand struct {
	Path    string
	Content string
}

func (c *CreateFileCommand) Execute() error {
	dir := filepath.Dir(c.Path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	return os.WriteFile(c.Path, []byte(c.Content), 0644)
}

func (c *CreateFileCommand) Undo() error {
	return os.Remove(c.Path)
}

func (c *CreateFileCommand) Description() string {
	return "ファイル作成: " + c.Path
}

// DeleteFileCommand deletes a file, saving its content for undo.
type DeleteFileCommand struct {
	Path           string
	savedContent   []byte
	savedFileExist bool
}

func (d *DeleteFileCommand) Execute() error {
	data, err := os.ReadFile(d.Path)
	if err != nil {
		return err
	}
	d.savedContent = data
	d.savedFileExist = true
	return os.Remove(d.Path)
}

func (d *DeleteFileCommand) Undo() error {
	if !d.savedFileExist {
		return nil
	}
	dir := filepath.Dir(d.Path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	return os.WriteFile(d.Path, d.savedContent, 0644)
}

func (d *DeleteFileCommand) Description() string {
	return "ファイル削除: " + d.Path
}

// CompositeCommand executes multiple commands as a group.
type CompositeCommand struct {
	Commands []Command
	executed []Command
}

func (cc *CompositeCommand) Execute() error {
	cc.executed = nil
	for _, cmd := range cc.Commands {
		if err := cmd.Execute(); err != nil {
			// Undo already executed commands on failure
			for i := len(cc.executed) - 1; i >= 0; i-- {
				_ = cc.executed[i].Undo()
			}
			return err
		}
		cc.executed = append(cc.executed, cmd)
	}
	return nil
}

func (cc *CompositeCommand) Undo() error {
	for i := len(cc.executed) - 1; i >= 0; i-- {
		if err := cc.executed[i].Undo(); err != nil {
			return err
		}
	}
	cc.executed = nil
	return nil
}

func (cc *CompositeCommand) Description() string {
	return "複合コマンド"
}

File: apps/php/design-pattern/src/Command/Command.php
<?php

declare(strict_types=1);

namespace DesignPattern\Command;

interface Command
{
    public function execute(): void;

    public function undo(): void;

    public function getDescription(): string;
}

class CreateFileCommand implements Command
{
    private string $path;
    private string $contents;

    public function __construct(string $path, string $contents = '')
    {
        $this->path = $path;
        $this->contents = $contents;
    }

    public function execute(): void
    {
        file_put_contents($this->path, $this->contents);
    }

    public function undo(): void
    {
        if (file_exists($this->path)) {
            unlink($this->path);
        }
    }

    public function getDescription(): string
    {
        return "ファイル作成: {$this->path}";
    }
}

class DeleteFileCommand implements Command
{
    private string $path;
    private string $savedContents = '';

    public function __construct(string $path)
    {
        $this->path = $path;
    }

    public function execute(): void
    {
        if (file_exists($this->path)) {
            $this->savedContents = file_get_contents($this->path);
            unlink($this->path);
        }
    }

    public function undo(): void
    {
        file_put_contents($this->path, $this->savedContents);
    }

    public function getDescription(): string
    {
        return "ファイル削除: {$this->path}";
    }
}

class CompositeCommand implements Command
{
    /** @var Command[] */
    private array $commands = [];
    private string $description;

    public function __construct(string $description = '複合コマンド')
    {
        $this->description = $description;
    }

    public function addCommand(Command $command): void
    {
        $this->commands[] = $command;
    }

    public function execute(): void
    {
        foreach ($this->commands as $command) {
            $command->execute();
        }
    }

    public function undo(): void
    {
        foreach (array_reverse($this->commands) as $command) {
            $command->undo();
        }
    }

    public function getDescription(): string
    {
        return $this->description;
    }

    /** @return Command[] */
    public function getCommands(): array
    {
        return $this->commands;
    }
}
